crypto-attributes/src/entity_derive.rs

```diff
@@ -0,0 +1,567 @@
+        match data {
+            Data::Struct(DataStruct {
+                fields: Fields::Named(named_fields),
+                ..
+            }) => Ok(named_fields),
+            _ => Err(syn::Error::new(span, "Expected a struct with named fields.")),
+        }
+    }
+}
+
+/// Less abstract version of [KeyStoreEntity] that has all the fields flattened
+/// ready for usage in `quote!()`.
+pub(super) struct KeyStoreEntityFlattened {
+    struct_name: Ident,
+    collection_name: String,
+    id: Ident,
+    id_name: String,
+    id_type: ColumnType,
+    all_columns: Vec<Ident>,
+    all_column_names: Vec<String>,
+    blob_columns: Vec<Ident>,
+    blob_column_names: Vec<String>,
+    no_upsert: bool,
+}
+
+impl KeyStoreEntityFlattened {
+    pub(super) fn entity_impl(&self) -> proc_macro2::TokenStream {
+        let entity_base_impl = self.entity_base_impl();
+        let entity_generic_impl = self.entity_generic_impl();
+        let entity_wasm_impl = self.entity_wasm_impl();
+        let entity_transaction_ext_impl = self.entity_transaction_ext_impl();
+
+        quote! {
+            #entity_base_impl
+            #entity_generic_impl
+            #entity_wasm_impl
+            #entity_transaction_ext_impl
+        }
+    }
+
+    fn entity_base_impl(&self) -> proc_macro2::TokenStream {
+        let Self {
+            collection_name,
+            struct_name,
+            ..
+        } = self;
+
+        // Identical for both wasm and non-wasm
+        quote! {
+            #[cfg_attr(target_family = "wasm", async_trait::async_trait(?Send))]
+            #[cfg_attr(not(target_family = "wasm"), async_trait::async_trait)]
+            impl crate::entities::EntityBase for #struct_name {
+                type ConnectionType = crate::connection::KeystoreDatabaseConnection;
+                type AutoGeneratedFields = ();
+                const COLLECTION_NAME: &'static str = #collection_name;
+
+                fn to_missing_key_err_kind() -> crate::MissingKeyErrorKind {
+                    crate::MissingKeyErrorKind::#struct_name
+                }
+
+                fn to_transaction_entity(self) -> crate::transaction::dynamic_dispatch::Entity {
+                    crate::transaction::dynamic_dispatch::Entity::#struct_name(self)
+                }
+            }
+        }
+    }
+
+    fn entity_generic_impl(&self) -> proc_macro2::TokenStream {
+        let Self {
+            collection_name,
+            struct_name,
+            id,
+            id_type,
+            id_name,
+            blob_columns,
+            blob_column_names,
+            all_columns,
+            ..
+        } = self;
+
+        let string_id_conversion = matches!(id_type, ColumnType::String).then(|| {
+            quote! { let #id: String = id.try_into()?; }
+        });
+
+        let id_getter = match id_type {
+            ColumnType::String => quote! {self.#id.as_bytes() },
+            ColumnType::Bytes => quote! { &self.#id[..] },
```

Comment:
It's equivalent, but IMO `self.#id.as_slice()` would be better. It introduces a nice parallel with the string case, and is easier to understand for people unfamiliar with the slice-everything idiom.